### PR TITLE
feature: Introduce `SpacesInsideParenthesesFixer`

### PR DESCRIPTION
**Pull request by SuperDJ**
Added new SpacesInsideParenthesisFixer in favor of NoSpacesInsideParenthesisFixer. The new fixer has a config for no spaces or with spaces in parenthesis.

Should fix https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3855 and should really close https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4706

# Custom rule
As I'm not sure sure when this is merged I created a [custom rule](https://packagist.org/packages/superdj/spaces-in-parentheses-php-cs-fixer) for the mean time.

### CODE REVIEW

**Comment by SuperDJ** [comment]
@SpacePossum. @keradus is there anything left to do before this is RTM?

**Comment by shakaran** [comment]
@SuperDJ Please, could you update the conflicting files or rebase? Thanks

File: src/Fixer/Whitespace/NoSpacesInsideParenthesisFixer.php
Comment (kubawerlos) [inline src/Fixer/Whitespace/NoSpacesInsideParenthesisFixer.php:24]: This is definitely not enough - please make it an `AbstractProxyFixer` to the new one (see `PhpdocNoAliasTagFixer` how it's done).

**Review by kubawerlos — DISMISSED** [review]

**Comment by SuperDJ** [comment]
@shakaran should be done now

File: doc/rules/whitespace/spaces_inside_parenthesis.rst
Comment (keradus) [inline doc/rules/whitespace/spaces_inside_parenthesis.rst:13]: is any community using `none` ?

File: tests/Fixer/Whitespace/SpacesInsideParenthesisFixerTest.php
Comment (keradus) [inline tests/Fixer/Whitespace/SpacesInsideParenthesisFixerTest.php:30]: pls import it

File: tests/Fixer/Whitespace/SpacesInsideParenthesisFixerTest.php
Comment (keradus) [inline tests/Fixer/Whitespace/SpacesInsideParenthesisFixerTest.php:321]:
```suggestion
multiply((2 + 3) * 4);
```
( so we see nested `)` was also covered)

File: tests/Fixer/Whitespace/SpacesInsideParenthesisFixerTest.php
Comment (keradus) [inline tests/Fixer/Whitespace/SpacesInsideParenthesisFixerTest.php:318]: do we have equivalent test for other configuration?

File: doc/rules/whitespace/spaces_inside_parenthesis.rst
Comment (SuperDJ) [reply doc/rules/whitespace/spaces_inside_parenthesis.rst:13]: @keradus setting it to `none` which is the default is also equivalent to setting `no_spaces_inside_parenthesis => true` in config

File: tests/Fixer/Whitespace/SpacesInsideParenthesisFixerTest.php
Comment (keradus) [inline tests/Fixer/Whitespace/SpacesInsideParenthesisFixerTest.php:86]: could be tested with main method as yet another item in provider

File: doc/rules/whitespace/spaces_inside_parenthesis.rst
Comment (keradus) [reply doc/rules/whitespace/spaces_inside_parenthesis.rst:13]: yes, you are right.
let me set the question properly:
is any community using `spaces`? i believe it's against PSR

Comment (SuperDJ) [reply doc/rules/whitespace/spaces_inside_parenthesis.rst:13]: @keradus as mentioned in https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4706 the WordPress community is using it

File: src/Fixer/Whitespace/NoSpacesInsideParenthesisFixer.php
Comment (julienfalque) [inline src/Fixer/Whitespace/NoSpacesInsideParenthesisFixer.php:28]: This fixer should be deprecated using `DeprecatedFixerInterface`.

**Comment by SpacePossum** [comment]
just wondering, is the style with `spaces` widely used?

**Comment by SuperDJ** [comment]
@SpacePossum as mentioned before (and in https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3855 and https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4706) it is used in the WordPress community.

**Comment by EarthmanWeb** [comment]
Eager to see this merged....

"Only those with write access to this repository can merge pull requests."

That's not me ;)

**Comment by SuperDJ** [comment]
@EarthmanWeb there is now a [custom rule](https://packagist.org/packages/superdj/spaces-in-parentheses-php-cs-fixer) for this

**Comment by SuperDJ** [comment]
@vena sorry for the late response. I found that the `NoSpacesInsideParenthesisFixer` has the same behaviour.

**Comment by EarthmanWeb** [comment]
> @vena sorry for the late response. I found that the `NoSpacesInsideParenthesisFixer` has the same behaviour.

Thanks for getting back to me, @SuperDJ 

I have found that the NoSpacesInsideParenthesisFixer
set to false will leave the spaces alone, but it will *not introduce new ones*

Have you found the same?

I will try the other rule that you posted

**Comment by sovetski** [comment]
Hi, any news about this pull request?..

**Review by kubawerlos — COMMENTED** [review]
Technically it looks fine, but still, WordPress community producing (click on your own risk, you have been warned) [such code](https://github.com/WordPress/WordPress/blob/6.1.1/wp-includes/class-wp.php#L416-L562) without a single test (to be precise: whole WordPress codebase does not have a single test) is not a community on which anything should be based, in terms of good practices or standards.

**Comment by EarthmanWeb** [comment]
@kubawerlos 

Testing wp core may be different than you expect, but there are indeed tests

https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/

**Comment by defunctl** [comment]
> Technically it looks fine, but still, WordPress community producing (click on your own risk, you have been warned) [such code](https://github.com/WordPress/WordPress/blob/6.1.1/wp-includes/class-wp.php#L416-L562) without a single test (to be precise: whole WordPress codebase does not have a single test) is not a community on which anything should be based, in terms of good practices or standards.

While WordPress's coding practices leave a lot to be desired (to put it as nicely as possible), the repo you linked is just their build/release repo.

The source repo is https://github.com/WordPress/wordpress-develop and the [tests are available here](https://github.com/WordPress/wordpress-develop/tree/trunk/tests).

At the end of the day, being able to enforce or not enforce spaces inside parenthesis for your own project would certainly be beneficial for many.

**Comment by Wirone** [comment]
@SuperDJ any chance for a rebase? It's good that you share your work through [custom fixer](https://github.com/SuperDJ/spaces-in-parentheses-php-cs-fixer), but I believe it would be beneficial to have it in the core. If you would like to work on it, please change the naming though (`parenthesis` → `parentheses`), as in your custom fixer.

**Comment by SuperDJ** [comment]
@Wirone I will only rebase if it's confirmed to be added after that. If not I will not spent more time on it.

**Comment by Wirone** [comment]
@SuperDJ fair enough 🙂. Personally I want to merge it, but I did not review it completely and I can't promise you that it will be merged as-is. Naming is something that I noticed when looking through the comments and your repo, but I did not dig deeper. However, @kubawerlos said it technically looks fine, so I believe it's almost ready 🙂.

**Comment by kubawerlos** [comment]
> I will only rebase if it's confirmed to be added after that. If not I will not spent more time on it.

I haven't seen blackmailing in this repo yet 😀

Please, keep in mind that everyone here is coding/maintaining for free. I'd leave the final decision to @keradus about merging/not merging this, as, after all, it's his name visible in the CLI.

**Comment by sovetski** [comment]
> I haven't seen blackmailing in this repo yet 😀

I don't think he said that in blackmail mode.

I understand him because his pull request is from almost 2 years ago, and as you said everyone in open source works for free, but if in the end we are working for something that will never be merged it must be frustrating 😕

File: src/Fixer/Whitespace/SpacesInsideParenthesisFixer.php
Comment (Wirone) [inline src/Fixer/Whitespace/SpacesInsideParenthesisFixer.php:1]: It should be named `SpacesInsideParenthesesFixer`, because `parentheses` is a valid plural form.

File: src/Fixer/Whitespace/SpacesInsideParenthesisFixer.php
Comment (Wirone) [inline src/Fixer/Whitespace/SpacesInsideParenthesisFixer.php:153]:
```suggestion
            (new FixerOptionBuilder('space', 'Whether to have `single` or `none` spaces inside parenthesis.'))
                ->setAllowedValues(['none', 'single'])
```

- `space: spaces` configuration looks weird
- it's not configurable, so it's actually one possible spacing (single space)

**Review by Wirone — CHANGES_REQUESTED** [review]
Basic feedback before digging deeper 🙂.

File: src/Fixer/Whitespace/SpacesInsideParenthesisFixer.php
Comment (SuperDJ) [reply src/Fixer/Whitespace/SpacesInsideParenthesisFixer.php:153]: I think `single` is clear enough. I used `spaces` more as a "working name".

**Comment by Wirone** [comment]
Daughter woke up during refactor and I missed some places, I'll fix it later 😉.

**Comment by Wirone** [comment]
At this point the only things missing:

- integration tests
- code review 😉 

The latter can be done already, please remember it's not my code and even I did not review it enough - I just fixed errors from CI and made some slight improvements.

**Comment by sovetski** [comment]
I feel like it's never gonna be merged 😭

**Review by kubawerlos — APPROVED** [review]

**Review by Wirone — APPROVED** [review]

**Comment by Wirone** [comment]
Thank you @SuperDJ 🍻

**Comment by SuperDJ** [comment]
@Wirone Thank you for all the additional work required to get this merged

**Pull request merged**
